File: ref/rust/src/segwit_addr.rs
// Copyright (c) 2017 Clark Moody
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

use bech32;

/// Witness `WitnessProgram` type
#[derive(Debug)]
pub struct WitnessProgram {
    /// Witness program version
    pub version: u8,
    /// Witness program content
    pub program: Vec<u8>
}

impl WitnessProgram {
    /// Converts a `WitnessProgram` to a script public key
    ///
    /// The format for the output is 
    /// `[version, program length, <program>]`
    pub fn to_scriptpubkey(&self) -> Vec<u8> {
        let mut pubkey: Vec<u8> = Vec::new();
        let mut v = self.version;
        if v > 0 {
            v += 0x80;
        }
        pubkey.push(v);
        pubkey.push(self.program.len() as u8);
        pubkey.extend_from_slice(&self.program);
        pubkey
    }

    pub fn from_scriptpubkey(pubkey: Vec<u8>) -> Result<WitnessProgram, ()> {
        // TODO: Validate according to BIP141
        
        // We need a version byte and a program length byte, with a program at 
        // least 2 bytes long.
        if pubkey.len() < 4 {
            return Err(())
        }
        let proglen: usize = pubkey[1] as usize;
        // Check that program length byte is consistent with pubkey length
        if pubkey.len() != 2 + proglen {
            return Err(())
        }
        // Process script version
        let mut v: u8 = pubkey[0];
        if v > 0x80 {
            v -= 0x80;
        }
        let (_, program) = pubkey.split_at(2);
        Ok(WitnessProgram {
            version: v,
            program: program.to_vec()
        })
    }

    pub fn clone(&self) -> WitnessProgram {
        WitnessProgram {
            version: self.version,
            program: self.program.to_vec()
        }
    }
}

pub fn encode(hrp: String, p: WitnessProgram) -> EncodeResult {
    let mut data: Vec<u8> = vec![p.version];
    // Convert 8-bit program into 5-bit
    let p5 = match convert_bits(p.program.to_vec(), 8, 5, true) {
        Ok(p) => p,
        Err(_) => {
            return Err(bech32::CodingError::InvalidChar)
        }
    };
    data.extend_from_slice(&p5);
    let enc_result = bech32::encode(
        bech32::Bech32 {hrp: hrp.clone(), data: data});
    if enc_result.is_err() {
        return Err(bech32::CodingError::InvalidChar)
    }
    let address = enc_result.unwrap();
    let dec_result = decode(hrp, address.clone());
    if dec_result.is_err() {
        return Err(bech32::CodingError::InvalidChar)
    }
    Ok(address)
}

/// Decodes a segwit address
///
/// Verifies that the `address` contains the expected human-readable part `hrp`
/// and decodes as proper Bech32-encoded string.
/// 
/// Returns the witness `WitnessProgram`
/// 
/// 
/// # Examples
/// 
pub fn decode(hrp: String, address: String) -> DecodeResult {
    let dec = bech32::decode(address);
    if dec.is_err() {
        return Err(dec.unwrap_err())
    }
    let b32 = dec.unwrap();
    if b32.hrp != hrp {
        return Err(bech32::CodingError::InvalidChar)
    }
    if b32.data.len() == 0 || b32.data.len() > 65 {
        return Err(bech32::CodingError::InvalidLength)
    }
    if b32.data[0] > 16 {
        // Invalid script version
        return Err(bech32::CodingError::InvalidChar)
    }
    let mut ret = WitnessProgram {
        version: b32.data[0],
        program: Vec::new()
    };
    // Get the 5-bit program
    let (_, p5) = b32.data.split_at(1);
    // Convert to 8-bit program
    let p8 = match convert_bits(p5.to_vec(), 5, 8, false) {
        Ok(p) => p,
        Err(_) => {
            return Err(bech32::CodingError::InvalidChar)
        },
    };
    if p8.len() < 2 || p8.len() > 40 {
        return Err(bech32::CodingError::InvalidLength)    
    }
    // Check proper script length
    if ret.version == 0 && p8.len() != 20 && p8.len() != 32 {
        return Err(bech32::CodingError::InvalidLength)    
    }
    ret.program = p8;
    Ok(ret)
}

/// Convert between bit sizes
///
/// # Panics
/// Function will panic if attempting to convert `from` or `to` a bit size that
/// is larger than 8 bits.
fn convert_bits(data: Vec<u8>, from: u32, to: u32, pad: bool) -> ConvertResult {
    if from > 8 || to > 8 {
        panic!("convert_bits `from` and `to` parameters greater than 8");
    }
    let mut acc: u32 = 0;
    let mut bits: u32 = 0;
    let mut ret: Vec<u8> = Vec::new();
    let maxv: u32 = (1<<to) - 1;
    for value in data {
        let v: u32 = value as u32;
        if (v >> from) != 0 {
            // Input value exceeds `from` bit size
            return Err(())
        }
        acc = (acc << from) | v;
        bits += from;
        while bits >= to {
            bits -= to;
            ret.push(((acc >> bits) & maxv) as u8);
        }
    }
    if pad {
        if bits > 0 {
            ret.push(((acc << (to - bits)) & maxv) as u8);
        }
    } else if bits >= from || ((acc << (to - bits)) & maxv) != 0 {
        return Err(())
    }
    Ok(ret)
}

type ConvertResult = Result<Vec<u8>, ()>;

type DecodeResult = Result<WitnessProgram, bech32::CodingError>;
type EncodeResult = bech32::EncodeResult;

File: ref/elixir/lib/segwit_addr.ex
# Copyright (c) 2017 Adán Sánchez de Pedro Crespo
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

defmodule SegwitAddr do
  use Bitwise

  @moduledoc ~S"""
  Encode and decode BIP-0173 compliant SegWit addresses.
  """

  @doc ~S"""
  Encode a SegWit address.

  ## Examples

      iex> SegwitAddr.encode("bc", "0014751e76e8199196d454941c45d1b3a323f1433bd6")
      "bc1qw508d6qejxtdg4y5r3zarvary0c5xw7kv8f3t4"

      iex> SegwitAddr.encode("bc", 0, [117, 30, 118, 232, 25, 145, 150, 212,
      ...> 84, 148, 28, 69, 209, 179, 163, 35, 241, 67, 59, 214])
      "bc1qw508d6qejxtdg4y5r3zarvary0c5xw7kv8f3t4"
  """
  @spec encode(String.t, integer, list(integer)) :: String.t
  def encode(hrp, version, program) when is_list(program) do
    Bech32.encode(hrp, [version] ++ convert_bits(program, 8, 5))
  end

  @spec encode(String.t, String.t) :: String.t
  def encode(hrp, program) when is_binary(program) do
    <<version, _size, program::binary>> = Base.decode16!(program, case: :mixed)
    program
      |> :binary.bin_to_list
      |> (&(encode(hrp, version, &1))).()
  end

  @doc ~S"""
  Decode a SegWit address.

  ## Examples

      iex> SegwitAddr.decode("bc1qw508d6qejxtdg4y5r3zarvary0c5xw7kv8f3t4")
      {:ok, {"bc", 0, [117, 30, 118, 232, 25, 145, 150, 212,
      84, 148, 28, 69, 209, 179, 163, 35, 241, 67, 59, 214]}}
  """
  @spec decode(String.t)
  :: {:ok, {pos_integer, list(integer)}} | {:error,  String.t}
  def decode(addr) do
    case Bech32.decode(addr) do
      {:ok, {hrp, data}} ->
        [version | encoded] = data
        program = convert_bits(encoded, 5, 8, false)
        {:ok, {hrp, version, program}}
      error -> error
    end
  end

  @doc ~S"""
  Encode a witness program into a hexadecimal ScriptPubKey.

  ## Examples

      iex> SegwitAddr.to_script_pub_key(0, [117, 30, 118, 232, 25, 145, 150,
      ...> 212, 84, 148, 28, 69, 209, 179, 163, 35, 241, 67, 59, 214])
      "0014751e76e8199196d454941c45d1b3a323f1433bd6"
  """
  @spec to_script_pub_key(pos_integer, list(integer)) :: String.t
  def to_script_pub_key(version, program) do
    [
      if version == 0 do 0 else version + 0x50 end,
      Enum.count(program) | program
    ]
      |> :binary.list_to_bin()
      |> Base.encode16(case: :lower)
  end

  # General power-of-2 base conversion.
  defp convert_bits(data, from, to, pad \\ true) do
    max_v = (1 <<< to) - 1
    if (Enum.find(data, fn (c) -> c < 0 || (c >>> from) != 0 end)) do
      nil
    else
      {acc, bits, ret} = Enum.reduce(
        data,
        {0, 0, []},
        fn (value, {acc, bits, ret}) ->
          acc = ((acc <<< from) ||| value)
          bits = bits + from
          {bits, ret} = convert_bits_loop(to, max_v, acc, bits, ret)
          {acc, bits, ret}
        end
      )
      if (pad && bits > 0) do
        ret ++ [(acc <<< (to - bits)) &&& max_v]
      else
        if (bits > from || ((acc <<< (to - bits)) &&& max_v) > 0) do
          nil
        else
          ret
        end
      end
    end
  end

  # Recursive version of the inner loop of the convert_bits function
  defp convert_bits_loop(to, max_v, acc, bits, ret) do
    if (bits >= to) do
      bits = bits - to
      ret = ret ++ [(acc >>> bits) &&& max_v]
      convert_bits_loop(to, max_v, acc, bits, ret)
    else
      {bits, ret}
    end
  end

end
